homeassistant/components/youtube/config_flow.py

```diff
@@ -0,0 +1,199 @@
+from typing import Any
+
+from google.oauth2.credentials import Credentials
+from googleapiclient.discovery import build
+from googleapiclient.http import HttpRequest
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry, OptionsFlowWithConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_TOKEN
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers import config_entry_oauth2_flow
+from homeassistant.helpers.selector import (
+    SelectOptionDict,
+    SelectSelector,
+    SelectSelectorConfig,
+)
+
+from . import AsyncConfigEntryAuth
+from .const import AUTH, CONF_CHANNELS, DEFAULT_ACCESS, DOMAIN
+
+
+class OAuth2FlowHandler(
+    config_entry_oauth2_flow.AbstractOAuth2FlowHandler, domain=DOMAIN
+):
+    """Config flow to handle Google OAuth2 authentication."""
+
+    _data: dict[str, Any] = {}
+
+    DOMAIN = DOMAIN
+
+    reauth_entry: ConfigEntry | None = None
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: ConfigEntry,
+    ) -> YouTubeOptionsFlowHandler:
+        """Get the options flow for this handler."""
+        return YouTubeOptionsFlowHandler(config_entry)
+
+    @property
+    def logger(self) -> logging.Logger:
+        """Return logger."""
+        return logging.getLogger(__name__)
+
+    @property
+    def extra_authorize_data(self) -> dict[str, Any]:
+        """Extra data that needs to be appended to the authorize url."""
+        return {
+            "scope": " ".join(DEFAULT_ACCESS),
+            # Add params to ensure we get back a refresh token
+            "access_type": "offline",
+            "prompt": "consent",
+        }
+
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
+        """Perform reauth upon an API authentication error."""
+        self.reauth_entry = self.hass.config_entries.async_get_entry(
+            self.context["entry_id"]
+        )
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Confirm reauth dialog."""
+        if user_input is None:
+            return self.async_show_form(step_id="reauth_confirm")
+        return await self.async_step_user()
+
+    async def async_oauth_create_entry(self, data: dict[str, Any]) -> FlowResult:
+        """Create an entry for the flow, or update existing entry."""
+
+        service = build(
+            "youtube",
+            "v3",
+            credentials=Credentials(data[CONF_TOKEN][CONF_ACCESS_TOKEN]),
+        )
+        # pylint: disable=no-member
+        own_channel_request: HttpRequest = service.channels().list(
+            part="snippet", mine=True
+        )
+        response = await self.hass.async_add_executor_job(own_channel_request.execute)
+        user_id = response["items"][0]["id"]
+        self._data = data
```

Comment:
I think there are a few things that can go wrong here that would make sense to handle, such as the credentials ended up being invalid or other transient failure with the API, or the response doesn't have what was expected .

There are currently error abort reason strings that do not appear to be used at the moment, so either way something needs to be resolved there.

Reply:
Eventually I got the tests working, instead of forcing a time update I forced the coordinator to update. 

I do have a hard time trying to figure what's needed for a correctly working oauth setup and what's part of reauth since all implementations with oauth already have reauth in them so I don't know.

Reply:
What would be a good way to handle these? Because just having a try except with Exception and then just abort with "unknown" also doesn't sound like the way to go

Reply:
Yeah, I'd say catch the exceptions that can happen.